pstoremgr/pstoremgr.go

```diff
@@ -221,13 +236,95 @@ func (pm *Manager) SavePeerstore(addrs []api.Multiaddr) {
 	}
 	defer f.Close()
 
-	for _, a := range addrs {
-		f.Write([]byte(fmt.Sprintf("%s\n", a.Value().String())))
+	for _, pinfo := range pinfos {
+		addrs, err := peerstore.InfoToP2pAddrs(&pinfo)
+		if err != nil {
+			logger.Warning(err)
+			continue
+		}
+		for _, a := range addrs {
+			f.Write([]byte(fmt.Sprintf("%s\n", a.String())))
+		}
 	}
 }
 
-// SavePeerstoreForPeers calls PeersAddresses and then saves the peerstore
+// SavePeerstoreForPeers calls PeerInfos and then saves the peerstore
 // file using the result.
 func (pm *Manager) SavePeerstoreForPeers(peers []peer.ID) {
-	pm.SavePeerstore(pm.PeersAddresses(peers))
+	pm.SavePeerstore(pm.PeerInfos(peers))
+}
+
+// Bootstrap attempts to get as much as count connected peers by selecting
+// randomly from those in the libp2p host peerstore. It returns the number
+// if peers it sucessfully connected to.
+func (pm *Manager) Bootstrap(count int) int {
+	knownPeers := pm.host.Peerstore().PeersWithAddrs()
+	toSort := &peerSort{
+		pinfos: peerstore.PeerInfos(pm.host.Peerstore(), knownPeers),
+		pstore: pm.host.Peerstore(),
+	}
+
+	// Sort from highest to lowest priority
+	sort.Sort(toSort)
+
+	pinfos := toSort.pinfos
+	lenKnown := len(pinfos)
+	totalConns := 0
+
+	// keep conecting while we have peers in the store
+	// and we have not reached count.
+	for i := 0; i < lenKnown && totalConns < count; i++ {
+		pinfo := pinfos[i]
+		ctx, cancel := context.WithTimeout(pm.ctx, ConnectTimeout)
+		defer cancel()
+
+		logger.Infof("connecting to %s", pinfo.ID)
+		err := pm.host.Connect(ctx, pinfo)
+		if err != nil {
+			logger.Warning(err)
+			pm.SetPriority(pinfo.ID, 9999)
```

Comment:
Say, we got an error while connecting to a trusted peer. This would send it down the list. Isn't it more important to stay connected to trusted peers?

Reply:
Hmm  yeah, I will move the code that SetsPriority=0 in crdt consensus to the `setup()` function. This will ensure that even it was demoted on bootstrapping, it will go up the list soon after. Right now, crdt is setting priorities even before the peers are imported and I'm not sure how well that works (and it's not correctly setting the TTL on the addresses anyway because there will be no addresses, so that's buggy).

Good catch.